docs/proposals/configmap-rollout.md

```diff
@@ -0,0 +1,181 @@
+by including the update as part of another update) losing the ability to
+rollback to previous configuration.
+
+This doc proposes a solution for removing the burden of ConfigMap management
+from users by providing a mechanism that will work out of the box without
+having to manually rollout new ReplicaSets.
+
+## Implementation
+
+### Proposed solution
+
+Users should only ever need to create and update a single ConfigMap "foo".
+```sh
+$ kubectl create configmap foo --from-file=config.yml
+configmap "foo" created
+```
+
+We need a way to link "foo" to the Deployment that is going to be triggered
+every time "foo" is updated.
+```sh
+$ kubectl set triggers deployment/bar configmap/foo
+deployment "bar" updated
+```
+
+A `{resource-0}.kubernetes.io/triggered-by: {resource-1}/{name-1}` annotation
+will be added in the resource we want to trigger automatic updates on. In the
+example above, a `deployment.kubernetes.io/triggered-by: configmap/foo`
+annotation should be added in Deployment "bar". We may want to extended this
+mechanism to support cross-namespace triggering since vendors may want to
+trigger on various resource updates not necessarily found in the same namespace
+as the resource that expects to be triggered.
+
+As today, the ConfigMap needs to be mounted as a volume in the Deployment.
+```sh
+$ kubectl set volume deployment/bar -t configmap --configmapname foo
+deployment "bar" updated
+```
+
+A new controller needs to watch for Deployments and ConfigMaps. Whenever it
+spots a Deployment with a `triggered-by` annotation and a mounted ConfigMap,
+it will create a copy of the ConfigMap and mount the copy in the Deployment.
+It should generally track Deployments with `triggered-by` annotations, 
+create new ConfigMap copies on updates of the original ConfigMap, and mount
+them back in the Deployment. Only Deployments should work for this controller
+for now but in the future the controller can be extended to facilitate config 
+management for DaemonSets, PetSets, and any other resource that may need it.
+
+Once the copied ConfigMap is mounted in the Deployment, the deployment
+controller needs to create the new ReplicaSet and set the owner reference
+for the copy to point to the ReplicaSet. In that way we ensure that copies
+of the original ConfigMap will be pruned by the garbage collector once their
+owner ReplicaSets are deleted. Certain copies of a ConfigMap will end up 
+being shared by more than one ReplicaSets.
+
+Additionally, the deployment controller should not create a new ReplicaSet
+for a Deployment "bar" that holds a `triggered-by` annotation to a ConfigMap
+"foo". Instead, it should wait until the new controller updates the Deployment
+template with the first copy of "foo". This ensures that the original ConfigMap
+copy will not land in a ReplicaSet, hence will not be subject to garbage
+collection.
+
+
+### Alternative solutions
+
+* Embed ConfigMap as a template in the Deployment.
```

Comment:
With respect to inlining, how are we going to solve multiple configs for multiple containers in a single Deployment? Multiple ConfigMapTemplates or the ConfigMapTemplate contains the config for every container? There are various patterns for mutlicontainer Deployments and furthermore some users already run embarassingly a lot of containers in single deployments (and we definitely cannot stop them from doing so). Config management for them would be a nightmare if we inlined everything.

I have been thiking about a separate object since the conception of this proposal but the bar for new API objects is pretty high at this point. My main concern is that we may end up with a solution for Deployments and down the road PetSets and DaemonSets will come up with different behavior for handling configuration updates. That's why I picked a new controller and an annotation instead of introducing new API, even though the annotation is essentially new API. It is much easier to deprecate though in case we decide we want something else before Deployments are ready for GA.

That being said, if we can't agree on a new controller, or inlining, or a new API object, I would prefer if we just handled the ConfigMap garbage collection in this proposal.

cc: @bprashanth @mikedanese @smarterclayton @bgrant0607 for opinions